### ISSUE TITLE
TimeDistributed Layer Applied to Masked Layer (Functional API) not working under TF 2.2+

### ISSUE THREAD

**Issue by weidler** [tensorflow/tensorflow#45371]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 20.04
- TensorFlow installed from (source or binary): pip
- TensorFlow version (use command below): 2.3.1
- Python version: 3.8.5
- CUDA/cuDNN version: 10.1
- GPU model and memory: GForce GTX 1060 (6BG)

**Describe the current behavior**
The following implementation of a model using the Keras functional API worked under TF2.0 and TF2.1 (reduced example from a bigger project, reproduces error though)

```python
import tensorflow as tf
from tensorflow.python.keras.layers import TimeDistributed

shared = False
bs = 1
layer_sizes = (64, 64)

state_dimensionality, n_actions = 8, 4
rnn_choice = tf.keras.layers.SimpleRNN

sequence_length = None
inputs = tf.keras.Input(shape=(sequence_length, state_dimensionality,), batch_size=bs)
masked = tf.keras.layers.Masking()(inputs)

# build encoder
encoder_inputs = tf.keras.Input(shape=state_dimensionality, batch_size=bs)
encoder_x = encoder_inputs
for i in range(len(layer_sizes)):
    encoder_x = tf.keras.layers.Dense(layer_sizes[i],
                                      kernel_initializer=tf.keras.initializers.Orthogonal(gain=tf.sqrt(2.0)),
                                      bias_initializer=tf.constant_initializer(0.0))(encoder_x)
    encoder_x = tf.keras.layers.Activation("tanh")(encoder_x)

encoder = tf.keras.Model(inputs=encoder_inputs, outputs=encoder_x, name="policy_encoder")

# policy network; stateful, so batch size needs to be known
x = TimeDistributed(encoder, name="TD_policy")(masked)
x.set_shape([bs] + x.shape[1:])

x, *_ = rnn_choice(layer_sizes[-1],
                   stateful=True,
                   return_sequences=True,
                   return_state=True,
                   batch_size=bs,
                   name="policy_recurrent_layer")(x)

means = tf.keras.layers.Dense(n_actions, name="means",
                              kernel_initializer=tf.keras.initializers.Orthogonal(0.01),
                              bias_initializer=tf.keras.initializers.Constant(0.0))(x)

policy = tf.keras.Model(inputs=inputs, outputs=means, name="simple_rnn_policy")
```

However when I run this under TF2.2.0+ I receive the following error: 
`ValueError: Input tensor 'policy_recurrent_layer/zeros_like:0' enters the loop with shape (1, 64), but has shape (None, 64) after one iteration. To allow the shape to vary across iterations, use the `shape_invariants` argument of tf.while_loop to specify a less-specific shape.`

I can prevent this error in the following ways:
- not using `stateful=True` and removing the line `x.set_shape([bs] + x.shape[1:])`
- removing the Masking layer
- setting `bs=2` or higher

Sadly, none of these options would work with the remainder of our project.

**Describe the expected behavior**
Given that this was working under TF2.0 I would also expect this to work under 2.3.1 or at least to throw a useful error message hinting towards a breaking change.

**Comment by Saduf2019** [comment]
i am able to replicate the issue reported, please find the [gist here](https://colab.research.google.com/gist/Saduf2019/69427352b1acdf76e61b4a9bf1dddc56/untitled477.ipynb)

**Comment by weidler** [comment]
This remains to be an issue in TF2.4.0.

**Comment by weidler** [comment]
Is this being worked on? Or can someone recommend a workaround?

**Comment by qlzh727** [comment]
Sorry for the very late reply. Keras team is heavily under staffed at the moment and this issue hasn't been prioritized for now. We will take a closer look when we have time. Meanwhile, contribution from community are welcome.

**Comment by weidler** [comment]
Thank you for the reply @qlzh727. I do of course understand that you have to set priorities. 
I will try to find a solution myself then. If I find one, I will post it here.

**Comment by saikumarchalla** [comment]
Was able to to reproduce the issue using TF 2.5. Please find the gist [here](https://colab.research.google.com/gist/saikumarchalla/aab50160cd5c8a0a4aee4f8f9221f2dd/untitled477.ipynb).Thanks!

**Comment by pjpratik** [comment]
@weidler The issue seem to be resolved in TF 2.11. Could you please check this [gist](https://colab.research.google.com/gist/pjpratik/8663b0b533090df8e137af014477e10d/45371.ipynb) and let us know if it helps. Thank you
